Trimming in variable substitution (#4287)
moby/buildkit: frontend/dockerfile/docs/reference.md

  will be that value. If `variable` is not set then `word` will be the result.
- `${variable:+word}` indicates that if `variable` is set then `word` will be
  the result, otherwise the result is the empty string.
- `${variable#pattern}` will return the value of `variable` with the shortest
  expansion of the pattern stripped from the beginning

> - `${variable##pattern}` will return the value of `varaible` with the *longest*

Reviewer: varaible

Author: Fixed

- `${variable:+word}` indicates that if `variable` is set then `word` will be
  the result, otherwise the result is the empty string.
- `${variable#pattern}` will return the value of `variable` with the shortest
  expansion of the pattern stripped from the beginning
- `${variable##pattern}` will return the value of `variable` with the *longest*
  expansion of the pattern stripped from the beginning,
  e.g. for `sheep=baaah` the expression `${sheep#b*a}` will yield `h` rather
  than `aah`.
- `${variable%pattern}` and `${variable%%pattern}` will strip from the end,
  respectively.

In all cases, `word` can be any string, including additional environment
variables.

`pattern` can be a glob pattern where `?` will match any single character
and `*` any number of characters (including zero). Literal `?`s and `*`s can
be escaped with `\`.

Escaping is possible by adding a `\` before the variable: `\$foo` or `\${foo}`,
for example, will translate to `$foo` and `${foo}` literals respectively.

Example (parsed representation is displayed after the `#`):

```dockerfile
ARG IMAGE=busybox:latest
# FROM busybox:stable

> FROM ${IMAGE%:*}:stable

Reviewer: I guess if we change this example, we need to add `#syntax` with the minimum required version as well.

Reviewer 2: Will we make a new labs release with this one? -> `# syntax=docker/dockerfile:1.7-labs`

Reviewer: Unless anyone objects, then I think this goes directly to the stable release, as there isn't a whole new command/flag.

Author: Should I add the `# syntax=docker/dockerfile:1.7` then?

Reviewer: We should make it clear what syntax requires 1.7 and that the first examples do not. As there is only one code example here, then maybe it doesn't make sense to change it and leave it as is. 

@dvdksn Are you updating docs from master or from the release branch? If from master then we need to say `docker/dockerfile-upstream:master` until there is a release. Otherwise `dockerfile:1.7`.

Reviewer 2: I'm using master, so let's do docker/dockerfile-upstream:master

Author: Added, needs to be squashed before the merge.

Author: Reverted by @tonistiigi because it's not yet supported outside the master image.

- `${variable:+word}` indicates that if `variable` is set then `word` will be
  the result, otherwise the result is the empty string.

The following variable replacements are supported in a pre-release version of

> Dockerfile syntax when defining `#syntax=docker/dockerfile-upstream:master` on top
> of your Dockerfile:

Reviewer 2: nit

```suggestion
Dockerfile syntax, when using the `# syntax=docker/dockerfile-upstream:master` syntax
directive in your Dockerfile:
```

Example (parsed representation is displayed after the `#`):

```dockerfile

> #syntax=docker/dockerfile-upstream:master

Reviewer 2: nit

```suggestion
# syntax=docker/dockerfile-upstream:master
```
